// testarmada-nightwatch-extra/lib/commands/mobile/clickMobileEl.js
"use strict";

var _util = require("util");

var _util2 = _interopRequireDefault(_util);

var _baseMobileCommand = require("../../base-mobile-command");

var _baseMobileCommand2 = _interopRequireDefault(_baseMobileCommand);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ClickMobileEl = function ClickMobileEl() {
  var nightwatch = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : null;

  _baseMobileCommand2.default.call(this, nightwatch);
  this.cmd = "clickmobileel";
};

_util2.default.inherits(ClickMobileEl, _baseMobileCommand2.default);

ClickMobileEl.prototype.do = function (value) {
  var self = this;

  this.client.api.elementIdClick(value.ELEMENT, function (result) {
    if (result.status === 0) {
      self.pass(result.value);
    } else {
      var errorMsg = null;
      if (result.error) {
        if (result.error.indexOf("not visible") > -1) {
          errorMsg = self.failureMessage + "[SELECTOR_NOT_VISIBLE]";
        } else {
          errorMsg = self.failureMessage + "[" + result.error + "]";
        }
      }
      self.fail(null, null, errorMsg);
    }
  });
};

ClickMobileEl.prototype.command = function (using, selector, cb) {
  this.selector = selector;
  this.using = using;
  this.cb = cb;

  this.successMessage = "Selector '" + this.using + ":" + this.selector + "' " + "was clicked after %d milliseconds.";
  this.failureMessage = "Selector '" + this.using + ":" + this.selector + "' " + "was not clicked after %d milliseconds.";

  this.startTime = new Date().getTime();

  this.seenCount = 0;
  this.checkConditions();

  return this;
};

module.exports = ClickMobileEl;
