// testarmada-nightwatch-extra/lib/commands/mobile/getMobileEl.js
"use strict";

var _util = require("util");

var _util2 = _interopRequireDefault(_util);

var _baseMobileCommand = require("../../base-mobile-command");

var _baseMobileCommand2 = _interopRequireDefault(_baseMobileCommand);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var GetMobileEl = function GetMobileEl() {
  var nightwatch = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : null;

  _baseMobileCommand2.default.call(this, nightwatch);
  this.cmd = "getmobileel";
};

_util2.default.inherits(GetMobileEl, _baseMobileCommand2.default);

GetMobileEl.prototype.do = function (value) {
  this.pass(value);
};

GetMobileEl.prototype.command = function (using, selector, cb) {
  this.selector = selector;
  this.using = using;
  this.cb = cb;

  this.successMessage = "Selector '" + this.using + ":" + this.selector + "' " + "was visible after %d milliseconds.";
  this.failureMessage = "Selector '" + this.using + ":" + this.selector + "' " + "was not visible after %d milliseconds.";

  this.startTime = new Date().getTime();

  // Track how many times selector is successfully checked by /element protocol
  this.seenCount = 0;
  this.checkConditions();

  return this;
};

module.exports = GetMobileEl;
