// testarmada-nightwatch-extra/lib/commands/mobile/setMobileElValue.js
"use strict";

var _util = require("util");

var _util2 = _interopRequireDefault(_util);

var _baseMobileCommand = require("../../base-mobile-command");

var _baseMobileCommand2 = _interopRequireDefault(_baseMobileCommand);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var SetMobileElValue = function SetMobileElValue() {
  var nightwatch = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : null;

  _baseMobileCommand2.default.call(this, nightwatch);
  this.cmd = "setmobileelvalue";
};

_util2.default.inherits(SetMobileElValue, _baseMobileCommand2.default);

SetMobileElValue.prototype.do = function (value) {
  var self = this;

  this.client.api.elementIdValue(value.ELEMENT, this.valueToSet, function (result) {
    if (result.status === 0) {
      self.pass(result.value);
    } else {
      self.fail(null, null, result.error);
    }
  });
};

/*eslint max-params:["error", 4] */
SetMobileElValue.prototype.command = function (using, selector, valueToSet, cb) {
  this.selector = selector;
  this.using = using;
  this.valueToSet = valueToSet;

  this.cb = cb;

  this.successMessage = "Selector '" + this.using + ":" + this.selector + "' " + ("set value to '" + this.valueToSet + "' after %d milliseconds.");
  this.failureMessage = "Selector '" + this.using + ":" + this.selector + "' " + ("can't set value to '" + this.valueToSet + "' after %d milliseconds.");

  this.startTime = new Date().getTime();

  this.seenCount = 0;
  this.checkConditions();

  return this;
};

module.exports = SetMobileElValue;
