// testarmada-nightwatch-extra/lib/commands/clickEl.js
"use strict";

var _util = require("util");

var _util2 = _interopRequireDefault(_util);

var _selector = require("../util/selector");

var _selector2 = _interopRequireDefault(_selector);

var _baseCommand = require("../base-command");

var _baseCommand2 = _interopRequireDefault(_baseCommand);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ClickEl = function ClickEl() {
  var nightwatch = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : null;
  var customizedSettings = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : null;

  _baseCommand2.default.call(this, nightwatch, customizedSettings);
  this.cmd = "clickel";
};

_util2.default.inherits(ClickEl, _baseCommand2.default);

ClickEl.prototype.do = function (magellanSel) {
  var self = this;
  var now = new Date().getTime();
  this.time.executeAsyncTime = now - self.startTime;

  this.client.api.click("css selector", "[" + this.selectorPrefix + "='" + magellanSel + "']", function () {
    self.time.seleniumCallTime = new Date().getTime() - now;
    self.pass();
  });
};

/*eslint no-unused-vars:0 */
ClickEl.prototype.injectedJsCommand = function ($el) {
  return "return $el[0].getAttribute('data-magellan-temp-automation-id')";
};

ClickEl.prototype.command = function (selector, cb) {
  this.selector = _selector2.default.normalize(selector);
  this.cb = cb;

  this.successMessage = "Selector <" + this.selector + "> clicked after %d milliseconds";
  this.failureMessage = "Selector <" + this.selector + "> could not be clicked after %d milliseconds";

  this.startTime = new Date().getTime();

  // Track how many times we've seen selector as :visible
  this.seenCount = 0;
  this.decide();
  this.checkConditions();

  return this;
};

module.exports = ClickEl;
