// testarmada-nightwatch-extra/lib/commands/getElValue.js
"use strict";

var _util = require("util");

var _util2 = _interopRequireDefault(_util);

var _selector = require("../util/selector");

var _selector2 = _interopRequireDefault(_selector);

var _baseCommand = require("../base-command");

var _baseCommand2 = _interopRequireDefault(_baseCommand);

var _stats = require("../util/stats");

var _stats2 = _interopRequireDefault(_stats);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var GetElValue = function GetElValue() {
  var nightwatch = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : null;
  var customizedSettings = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : null;

  _baseCommand2.default.call(this, nightwatch, customizedSettings);
  this.cmd = "getelvalue";
};

_util2.default.inherits(GetElValue, _baseCommand2.default);

GetElValue.prototype.do = function (value) {
  this.pass(value);
};

/*eslint no-unused-vars:0 */
GetElValue.prototype.injectedJsCommand = function ($el, sizzle) {
  return "return sizzle.getText($el)";
};

GetElValue.prototype.pass = function (actual) {
  this.time.totalTime = new Date().getTime() - this.startTime;
  this.client.assertion(true, actual, actual, _util2.default.format(this.successMessage, this.time.totalTime), true);

  (0, _stats2.default)({
    capabilities: this.client.options.desiredCapabilities,
    type: "command",
    cmd: this.cmd,
    value: this.time
  });

  if (this.cb) {
    this.cb.apply(this.client.api, [actual, this.selector]);
  }
  this.emit("complete");
};

GetElValue.prototype.command = function (selector, cb) {
  this.selector = _selector2.default.normalize(selector);
  this.cb = cb;

  this.successMessage = "Selector '" + this.selector + "' was visible after %d milliseconds.";
  this.failureMessage = "Selector '" + this.selector + "' was not visible after %d milliseconds.";

  this.startTime = new Date().getTime();

  // Track how many times we've seen selector as :visible
  this.seenCount = 0;
  this.decide();
  this.checkConditions();

  return this;
};

module.exports = GetElValue;
