// testarmada-nightwatch-extra/lib/util/stats.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

exports.default = function (event) {
  var sdclient = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : null;
  var customizedOptions = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : null;

  /**
   *  event: {
            "capabilities": {
                id: 'chrome_latest_Windows_10_Desktop',
                browserName: 'chrome',
                version: '50',
                platform: 'Windows 10',
                'tunnel-identifier': 'some-tunnel-value',
                name: 'Google'
            }
            "type"" "command"
            "cmd": "clickEl",
            "value": {
              totalTime: 300,
              seleniumCallTime: 180,
              executeAsyncTime: 120
            }
        }
   */

  var isEnabled = customizedOptions ? customizedOptions.isEnabled : ISENABLED;
  var host = customizedOptions ? customizedOptions.host : HOST;
  var port = customizedOptions ? customizedOptions.port : PORT;
  var prefix = customizedOptions ? customizedOptions.prefix : PREFIX;
  var jobName = customizedOptions ? customizedOptions.jobName : JOBNAME;

  if (!isEnabled || !host || !port) {
    return;
  }

  var fullEventKey = _util2.default.format("%s%s.%s.%s.%s.%s", prefix, event.capabilities.id || event.capabilities.browserName, event.capabilities["tunnel-identifier"] ? "tunnel" : "notunnel", jobName, event.type, event.cmd);

  var client = sdclient ? sdclient : new _statsdClient2.default({ host: host, port: port, prefix: prefix });

  client.timing(fullEventKey + ".totalTime", event.value.totalTime);
  client.timing(fullEventKey + ".seleniumCallTime", event.value.seleniumCallTime);
  client.timing(fullEventKey + ".executeAsyncTime", event.value.executeAsyncTime);
};

var _util = require("util");

var _util2 = _interopRequireDefault(_util);

var _statsdClient = require("statsd-client");

var _statsdClient2 = _interopRequireDefault(_statsdClient);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// second check for magellan compatibility
var ISENABLED = process.env.MAGELLAN_STATSD ? true : false;
var HOST = process.env.MAGELLANSTATS_STATSD_URL;
var PORT = process.env.MAGELLANSTATS_STATSD_PORT;
var PREFIX = process.env.MAGELLANSTATS_STATSD_PREFIX + ".";
var JOBNAME = process.env.MAGELLAN_JOB_NAME;

/*eslint complexity: ["error", 12]*/
