// testarmada-nightwatch-extra/lib/assertions/elContainsText.js
"use strict";

var _util = require("util");

var _util2 = _interopRequireDefault(_util);

var _selector = require("../util/selector");

var _selector2 = _interopRequireDefault(_selector);

var _baseAssertion = require("../base-assertion");

var _baseAssertion2 = _interopRequireDefault(_baseAssertion);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ElContainsText = function ElContainsText() {
  var nightwatch = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : null;
  var customizedSettings = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : null;

  _baseAssertion2.default.call(this, nightwatch, customizedSettings);
  this.cmd = "elcontainstext";
};

_util2.default.inherits(ElContainsText, _baseAssertion2.default);

ElContainsText.prototype.assert = function (actual, expected) {
  var pactual = actual.replace(/[\s|\n]+/g, " ");

  if (expected === undefined || pactual.indexOf(expected) < 0 && !new RegExp(expected).exec(pactual)) {
    this.fail(pactual, expected, this.message, this.failureDetails);
  } else {
    this.pass(pactual, expected, this.message);
  }
};

/* eslint-disable */
ElContainsText.prototype.injectedJsCommand = function ($el, sizzle) {
  return "return sizzle.getText($el)";
};

ElContainsText.prototype.command = function (selector, expected) {
  this.selector = _selector2.default.normalize(selector);
  this.expected = expected;

  this.message = _util2.default.format("Testing if selector <%s> contains text <%s> after %d milliseconds ", this.selector, this.expected);
  this.failureDetails = "actual result:[ %s ], expected:[ " + this.expected + " ]";
  this.notVisibleFailureMessage = "Selector '" + this.selector + "' was not visible after %d milliseconds.";

  this.startTime = new Date().getTime();

  // Track how many times we've seen selector as :visible
  this.seenCount = 0;
  this.decide();
  this.checkConditions();

  return this;
};

module.exports = ElContainsText;
